pkg/auth/auth.go

```diff
@@ -0,0 +1,173 @@
+package auth
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"path"
+	"strings"
+
+	logging "github.com/op/go-logging"
+)
+
+// ConfigEntry - Configuration for authentication
+type ConfigEntry struct {
+	Type    string `yaml:"type"`
+	Enabled bool   `yaml:"enabled"`
+}
+
+// Provider - an auth provider is an adapter that provides the principal
+// object required for authentication. This can be a User, a System, or some
+// other entity.
+type Provider interface {
+	GetPrincipal(*http.Request) (Principal, error)
+}
+
+// Principal - principal user or other identity of some kind with access to the
+// broker.
+type Principal interface {
+	GetType() string
+	GetName() string
+	// TODO: add roles?
+}
+
+// UserServiceAdapter - is the interface for a service that stores Users. It can
+// be anything you want: file, database, whatever as long as you can search and
+// validate them.
+type UserServiceAdapter interface {
+	FindByLogin(string) (User, error)
+	ValidateUser(string, string) bool
+}
+
+// User - a User from the service adapter
+type User struct {
+	Username string
+	Password string
+}
+
+// GetType - returns the type of Principal. This is a user principal.
+func (u User) GetType() string {
+	return "user"
+}
+
+// GetName - returns the name of the User Principal.
+func (u User) GetName() string {
+	return u.Username
+}
+
+// FileUserServiceAdapter - a file based UserServiceAdapter which seeds its
+// users from a file.
+type FileUserServiceAdapter struct {
```

Comment:
I prefer to keep the methods of the struct near the struct definition. What I did do was move the NewFileUser ctor method further down so that the structs methods are below the struct.